### PR TITLE
add ability to search developers by their bio or hero

### PR DESCRIPTION
**Pull request by themudassarhassan**
This PR is aiming to add a search field on `/developers` page to allow users to filter developers by their bio or hero. I have used the `pg_search` gem which implements the full text search feature of PostgresSQL using `to_tsvector` data type. Although this approach will work fine now but it will eventually get slow when we have hundreds of developer profiles so we also need to add index for that. I'll be creating a separate issue to add more details regarding its implementation.
### Pull request checklist

- [x] Your code contains tests relevant for code you modified
- [x] You have linted and tested the project with `bin/check`

### CODE REVIEW

File: app/components/developer_query_component.html.erb
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:9]: Is it possible to not have an `else` statement on the feature flag? Even if that means moving Sort all the way to the right even before this feature is enabled.

File: app/models/feature.rb
Comment (joemasilotti) [inline app/models/feature.rb:10]:
```suggestion
    when :search_filter
```

**Review by joemasilotti — COMMENTED** [review]
I know this isn't ready for review yet, but I wanted to get some comments in early to hopefully skip a round of review.

File: app/components/developer_query_component.html.erb
Comment (themudassarhassan) [reply app/components/developer_query_component.html.erb:9]: Yeah the Sort button will be moved to the right side if we don't add the `else` here. So do you suggest to remove the else statement?

Comment (joemasilotti) [reply app/components/developer_query_component.html.erb]: Fine by me! The search will eventually be ready for everyone, eventually.

**Comment by joemasilotti** [comment]
Hey @themudassarhassan, is this ready for review?

**Comment by themudassarhassan** [comment]
@joemasilotti I left a question for you on the PR I think you have missed it.

**Comment by joemasilotti** [comment]
Weird, I'm not seeing it! Can you link me or point to the code here?

**Comment by themudassarhassan** [comment]
Here is the [link](https://github.com/joemasilotti/railsdevs.com/pull/341/files#diff-cc13964906c303ef549a8fa3e5577bdef84b0c9563dd248f6d84cb83663b25ffR7). Pardon me if I didn't get it right.

**Comment by joemasilotti** [comment]
Weird, I'm not seeing anything. Did you do a review and not submit it? Either way, want to copy-paste the question here?

image

File: app/components/developer_query_component.html.erb
Comment (joemasilotti) [reply app/components/developer_query_component.html.erb:7]: Hmm, good point. I'd like it to "work" with and without the feature flag enabled.

What if we always write the `justify-end` class but only _add in_ the `justify-between` one if the feature is enabled? That should overwrite the first one, right?

Comment (joemasilotti) [reply app/components/developer_query_component.html.erb:7]: Or, if that feels/looks weird, what you have should be fine. We could pull it out to its own line of code, if need be.

Comment (themudassarhassan) [reply app/components/developer_query_component.html.erb:7]: Yeah, you are right, adding the `justify-between` class conditionally would override the `justify-end` class. Should I implement it that way?

Comment (joemasilotti) [reply app/components/developer_query_component.html.erb:7]: Yes please!

**Comment by themudassarhassan** [comment]
I have pushed the changes. Ready for review and merge 🚀

File: app/components/developer_query_component.html.erb
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:154]:
```suggestion
                      <%= form.text_field :search_query, value: search_query, class: "shadow-sm focus:ring-gray-500 focus:border-gray-500 block w-full sm:text-sm border-gray-300 rounded-md" %>
```

File: test/queries/developer_query_test.rb
Comment (joemasilotti) [inline test/queries/developer_query_test.rb:136]: Let's add some query here so it tests the value is getting passed along correctly.

File: app/models/developer.rb
Comment (themudassarhassan) [reply app/models/developer.rb:33]: I'm not sure if `pg_search` docs recommend it or not (they have some information on how to implement the index to speed up the search process, I'll be creating an issue with details once this PR merge) but the official documentation of Postgres does.
Screenshot 2022-03-23 at 10 22 49 PM

Comment (joemasilotti) [reply app/models/developer.rb:33]: Excellent, thanks for the link! I'll create an issue and assign it to you for indexes.

Comment (themudassarhassan) [reply app/models/developer.rb:33]: Sure sure 🚀

Comment (joemasilotti) [reply app/models/developer.rb:33]: If you comment on this issue I can assign it to you.

https://github.com/joemasilotti/railsdevs.com/issues/347

**Comment by themudassarhassan** [comment]
Hey, @joemasilotti I have pushed all feedback changes except one which is adding a search button next to the search field, regarding which I need some input. This is the UI on tablet layout, search input, button, and filters are quite close on this layout as shown in the image below:
Screenshot 2022-03-24 at 9 02 09 PM

We cannot add space because there isn't any available, what do you suggest?

**Comment by joemasilotti** [comment]
Ha, wow that's getting crowded. Let's leave off the submit button for now. I'm hoping to redesign these soon anyways. 

Is everything ready for review then?

**Comment by themudassarhassan** [comment]
Yes all other changes are done.

**Review by joemasilotti — APPROVED** [review]

**Comment by joemasilotti** [comment]
This looks great – thanks again for knocking this out! 💪

**Pull request merged**
